### ISSUE TITLE
support pytest

### ISSUE THREAD

**Issue by KotlinIsland** [DetachHead/basedpyright#311]
pylance:
```py
from pytest import fixture

@fixture()
def f() -> int:
    return 1


def test_asdf(f`: int`):
    reveal_type(f)  # Any
```

This is of course absolutely insane that it's `Any`.

I don't know the best way to handle this, via a plugin?

**Comment by DetachHead** [comment]
blocked by #22

i dont want to special-case specific modules if i can help it, especially 3rd party ones, so i don't really want to do this without a plugin system

**Comment by Daku24** [comment]
Is there a way to disable inlay hints ??
<img width="646" alt="image" src="https://github.com/DetachHead/basedpyright/assets/78242317/9a837b8a-0291-456c-84ef-f0d95634ea92">

**Comment by jorenham** [comment]
With pytest 8.3.2 and bpr 1.16.0 I get `Type of "f" is "int"` with

```python
from pytest import fixture
from typing_extensions import reveal_type


@fixture()
def f() -> int:
    return 1


def test_asdf(f: int):
    reveal_type(f)
```

so no `Any`more 🤷🏻‍♂️

**Comment by DetachHead** [comment]
sorry the example in the original post is a bit confusing. this part with the type annotation in backticks was just pseudocode we came up with to say that it should infer `f` as `int` and show it as an inlay hint:

```py
def test_asdf(f`: int`):
    reveal_type(f)
```

i updated that example to remove that part because i don't think this is really relevant to inlay hints. the inlay hint support is implied (as in, once we have pytest support and `f` gets inferred as `int`, the inlay hint will come naturally)

```py
@fixture()
def f() -> int:
    return 1


def test_asdf(f): # no type annotation should be needed, it should just infer it as int
    reveal_type(f) # int
```

**Comment by DetachHead** [comment]
once #336 is resolved we also want to disable `reportUnusedParameter` on pytest `hookspec`s (related: https://github.com/astral-sh/ruff/issues/9803)

**Comment by ewjoachim** [comment]
Something else that would be cool if supporting pytest fixures the way Pylance does it is the ability to cmd+click on a fixture to go to its definition

**Comment by jorenham** [comment]
> Something else that would be cool if supporting pytest fixures the way Pylance does it is the ability to cmd+click on a fixture to go to its definition

Yea, fixtures are pretty annoying to work with, and I think the lack of tooling support is one of the main reasons. So I also think that being able to jump to the fixture definitions like would make working with fixtures **a lot** less painful.

**Comment by filskas** [comment]
> Something else that would be cool if supporting pytest fixures the way Pylance does it is the ability to cmd+click on a fixture to go to its definition

@DetachHead should we create a separate issue for this? This makes working with tests quite painful

**Comment by DetachHead** [comment]
I would just consider that part of this issue

**Comment by filskas** [comment]
Could this be prioritized somehow? [Cursor relies on `basedpyright`](https://forum.cursor.com/t/cant-click-into-pytest-fixtures-ide-go-to-definition/128210) and the lack of this go-to-definition support has quite a big impact given the Cursor userbase.

**Comment by DetachHead** [comment]
i haven't had much time to work on basedpyright lately as i now have several other projects that i also maintain in my free time, as well as my full time job. i'm aware of cursor's large userbase and i'm glad basedpyright has gained widespread adoption, but i don't work for cursor.

anybody is welcome to submit a PR and i will review it when i have time (if you do, please take a look at [my comment regarding plugins](https://github.com/DetachHead/basedpyright/issues/22#issuecomment-3363785073) first).

otherwise [upvoting issues](https://github.com/DetachHead/basedpyright/issues/404) is the best way to help me (and other potential contributors) know what to prioritize.

**Comment by asukaminato0721** [comment]
> Could this be prioritized somehow? [Cursor relies on `basedpyright`](https://forum.cursor.com/t/cant-click-into-pytest-fixtures-ide-go-to-definition/128210) and the lack of this go-to-definition support has quite a big impact given the Cursor userbase.

I impl it here, so someone may able to port it.

https://github.com/facebook/pyrefly/pull/2075

**Comment by AmirTNinja** [comment]
Hey all,
is there any update on this issue or any ETA for fixing it? 

@DetachHead  is this possible to promote this request somehow? we would highly appreciate it

**Comment by benblank** [comment]
@AmirTNinja basedpyright uses [upvotes](https://docs.basedpyright.com/latest/development/contributing/#github-issues) to prioritize issues — you can vote by adding a 👍 reaction to any issue. Because the contributors only have so much time to spend on the project, the only other way to really promote an issue would be to convince a new contributor to donate some of their time to the project. 🙂

*Note: I don't speak for the project! I'm just another basedpyright user who stumbled on your comment.*

**Comment by asukaminato0721** [comment]
https://pyrefly.org/en/docs/pytest/ now supported here.
